File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private void hanoi(int start, int mid, int end, int n, List<int[]> answer) {
        if (n == 1) {
            answer.add(new int[] {start, end});
            return;
        }
        
        hanoi(start, end, mid, n - 1, answer);
        hanoi(start, mid, end, 1, answer);
        hanoi(mid, start, end, n - 1, answer);
    }
    
    public int[][] solution(int n) {
        List<int[]> answer = new ArrayList<>();
        hanoi(1, 2 ,3, n, answer);
        return answer.toArray(new int[0][]);
    }
}

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.py
def solution(n):
    answer = []
    hanoi(1, 2, 3, n, answer)
    return answer


def hanoi(start, mid, end, n, answer):
    if n == 1:
        answer.append([start, end])
        return 

    hanoi(start, end, mid, n - 1, answer)
    hanoi(start, mid, end, 1, answer)
    hanoi(mid, start, end, n - 1, answer)
